internal/builtinfunctions/functions_test.go

```diff
@@ -0,0 +1,600 @@
+		[]any{-0.1},
+		false,
+		0.0,
+	},
+	"floor-1": {
+		"floor",
+		[]any{0.0},
+		false,
+		0.0,
+	},
+	"floor-2": {
+		"floor",
+		[]any{1.1},
+		false,
+		1.0,
+	},
+	"floor-3": {
+		"floor",
+		[]any{2.9999},
+		false,
+		2.0,
+	},
+	"floor-4": {
+		"floor",
+		[]any{-0.1},
+		false,
+		-1.0,
+	},
+	"round-1": {
+		"round",
+		[]any{0.0},
+		false,
+		0.0,
+	},
+	"round-2": {
+		"round",
+		[]any{0.1},
+		false,
+		0.0,
+	},
+	"round-3": {
+		"round",
+		[]any{0.4},
+		false,
+		0.0,
+	},
+	"round-4": {
+		"round",
+		[]any{0.5},
+		false,
+		1.0,
+	},
+	"round-5": {
+		"round",
+		[]any{0.9},
+		false,
+		1.0,
+	},
+	"round-6": {
+		"round",
+		[]any{-0.9},
+		false,
+		-1.0,
+	},
+	"round-7": {
+		"round",
+		[]any{-0.5},
+		false,
+		-1.0,
+	},
+	"round-8": {
+		"round",
+		[]any{-0.01},
+		false,
+		0.0,
+	},
+	"abs-1": {
+		"abs",
+		[]any{0.0},
+		false,
+		0.0,
+	},
+	"abs-2": {
+		"abs",
+		[]any{-1.0},
+		false,
+		1.0,
+	},
+	"abs-3": {
+		"abs",
+		[]any{1.0},
+		false,
+		1.0,
+	},
+	"to-lower-1": {
+		"toLower",
+		[]any{""},
+		false,
+		"",
+	},
+	"to-lower-2": {
+		"toLower",
+		[]any{"abc"},
+		false,
+		"abc",
+	},
+	"to-lower-3": {
+		"toLower",
+		[]any{"ABC"},
+		false,
+		"abc",
+	},
+	"to-lower-4": {
+		"toLower",
+		[]any{"aBc"},
+		false,
+		"abc",
+	},
+	"to-upper-1": {
+		"toUpper",
+		[]any{""},
+		false,
+		"",
+	},
+	"to-upper-2": {
+		"toUpper",
+		[]any{"abc"},
+		false,
+		"ABC",
+	},
+	"to-upper-3": {
+		"toUpper",
+		[]any{"ABC"},
+		false,
+		"ABC",
+	},
+	"to-upper-4": {
+		"toUpper",
+		[]any{"aBc"},
+		false,
+		"ABC",
+	},
+	"split-string-1": {
+		"splitString",
+		[]any{"", ","},
+		false,
+		make([]string, 1), // The size appears to matter with DeepEquals
+	},
+	"split-string-2": {
+		"splitString",
+		[]any{"abc", ","},
+		false,
+		[]string{"abc"},
+	},
+	"split-string-3": {
+		"splitString",
+		[]any{"a,b", ","},
+		false,
+		[]string{"a", "b"},
+	},
+	"split-string-4": {
+		"splitString",
+		[]any{"a,b,c,d", ","},
+		false,
+		[]string{"a", "b", "c", "d"},
+	},
+	"split-string-5": {
+		"splitString",
+		[]any{"1,,2,", ","},
+		false,
+		[]string{"1", "", "2", ""},
+	},
+}
+
+func TestFunctionsBulk(t *testing.T) {
+	allFunctions := builtinfunctions.GetFunctions()
+	for name, tc := range testData {
+		testCase := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			functionToTest, funcFound := allFunctions[testCase.functionName]
+			if !funcFound {
+				t.Fatalf("Function %q not found", testCase.functionName)
+			}
+
+			output, err := functionToTest.Call(testCase.rawArguments)
+
+			if testCase.returnError && err == nil {
+				t.Fatalf("expected error in test case %q; error returned nil", name)
+			} else if !testCase.returnError && err != nil {
+				t.Fatalf("unexpected error in test case %q (%s)", name, err.Error())
+			}
+
+			if !reflect.DeepEqual(testCase.expectedResult, output) {
+				t.Fatalf("mismatch for test %q, expected: %v, got: %v", name, testCase.expectedResult, output)
+			}
```

Comment:
If returnError is true and we do get an error, we don't validate the error itself at the moment. Doing so would require adding another field to the struct used for the table of test cases. Yes, it basically just validates that it returns nil.

Reply:
You can use the `expectedResult` field -- what you compare it to depends on the `returnError` field -- so, we wouldn't need any new fields; we would just make full use of the fields that we've got.

Reply:
The expectedResult field is an any type. It would be possible to cast it to a string to do an assert contains. But I don't think that's necessary at the moment.

Reply:
The question is, is the contents of the `err` return part of the "interface" to this function?  What expectations should the caller of this function have with regard to the value returned in `err` under various circumstances?

If all we are claiming is that it is `nil` in the absence of an error and non-`nil` in the presence of an error, then this test as presented is sufficient.  But, I suspect that we want callers to be able to interrogate the `err` value and to be able to take appropriate actions (other than just to log it and die) depending on what it contains.

If that's the case, then we should be testing that `err` contains expected values (or, at least, values with specific characteristics) when we run specific scenarios.  Simply testing that it is non-`nil` isn't sufficient.

And, we can do that within this framework with some straightforward enhancements, by providing input for the assertion in the `expectedResult` field and by performing the appropriate check on that input when the `returnError` field is true.

Reply:
That is not worth it right now.